src/coreclr/vm/methodtable.cpp

```diff
@@ -2881,6 +2881,515 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
+                        else if (pFieldStart[1].GetSize() == 8)
+                            size |= STRUCT_SECOND_FIELD_SIZE_IS8;
+                        goto _End_arg;
+                    }
+                }
+                else if (fieldType == ELEMENT_TYPE_VALUETYPE)
+                {
+                    const NativeFieldDescriptor *pNativeFieldDescs = pMethodTable->GetNativeLayoutInfo()->GetNativeFieldDescriptors();
+                    NativeFieldCategory nfc = pNativeFieldDescs->GetCategory();
+
+                    if (nfc == NativeFieldCategory::NESTED)
+                    {
+                        MethodTable* pMethodTable2 = pNativeFieldDescs->GetNestedNativeMethodTable();
+
+                        if ((pMethodTable2->GetNumInstanceFieldBytes() > 8) || (pMethodTable2->GetNumIntroducedInstanceFields() > 1))
+                            goto _End_arg;
+                        size = GetLoongArch64PassStructInRegisterFlags((CORINFO_CLASS_HANDLE)pMethodTable2);
+                        if ((size & STRUCT_FLOAT_FIELD_ONLY_ONE) != 0)
+                        {
+                            if (pMethodTable2->GetNativeSize() == 8)
+                                size = STRUCT_FIRST_FIELD_DOUBLE;
+                            else
+                                size = STRUCT_FLOAT_FIELD_FIRST;
+                        }
+                        else if (pMethodTable2->GetNativeSize() == 8)
+                        {
+                            size = STRUCT_FIRST_FIELD_SIZE_IS8;
+                        }
+                        else
+                        {
+                            size = STRUCT_NO_FLOAT_FIELD;
+                            goto _End_arg;
+                        }
+                    }
+                    else if (nfc == NativeFieldCategory::FLOAT)
+                    {
+                        if (pFieldStart[0].GetSize() == 4)
+                        {
+                            size = STRUCT_FLOAT_FIELD_FIRST;
+                        }
+                        else if (pFieldStart[0].GetSize() == 8)
+                        {
+                            _ASSERTE(pMethodTable->GetNativeSize() == 8);
+                            size = STRUCT_FIRST_FIELD_DOUBLE;
+                        }
+                    }
+                }
+                else if (fieldType == ELEMENT_TYPE_CLASS)
+                {
+                    MethodTable* pMethodTable2  = pFieldStart->GetFieldTypeHandleThrowing().GetMethodTable();
+
+                    if ((pMethodTable2->GetNumInstanceFieldBytes() > 8) || (pMethodTable2->GetNumIntroducedInstanceFields() > 1))
+                    {
+                        size = STRUCT_NO_FLOAT_FIELD;
+                        goto _End_arg;
+                    }
+                    if (pMethodTable2->IsArray())
+                    {
+                        if ((pMethodTable2->GetNumInstanceFieldBytes() / pFieldStart[0].GetSize())  > 1)
```

Comment:
`GetNumInstanceFieldBytes` for arrays should be always zero. Is this path ever hit?

Reply:
> `GetNumInstanceFieldBytes` for arrays should be always zero. Is this path ever hit?

I remembered that the array cases hitted it. I use this is to get the count of the element of array.
I will check it now.

Reply:
> > `GetNumInstanceFieldBytes` for arrays should be always zero. Is this path ever hit?
> 
> I remembered that the array cases hitted it. I use this is to get the count of the element of array. I will check it now.

I have checked again liking
```
(gdb) p *pFieldStart
$2 = {m_pMTOfEnclosingClass = {static isRelative = false, m_ptr = 1097319343376}, m_mb = 151, m_isStatic = 0, m_isThreadLocal = 0, m_isRVA = 0, m_prot = 6,
  m_requiresFullMbValue = 0, m_dwOffset = 0, m_type = 18, m_debugName = 0xfff4116f18 "arr"}
(gdb) p *pMethodTable2    
$5 = {static MaxGenericParametersForSpecialMarkerType = 8, static s_pMethodDataCache = 0xaaaab93f30, static s_fUseParentMethodData = 1, static s_fUseMethodDataCache = 1,
  static UNION_MASK = 3,
  static c_DispatchMapSlotOffsets = 0xfff77d5f9e <MethodTable::c_DispatchMapSlotOffsets> "8@8H8@8H@HHP8@8H@HHP@HHPHPPXHHHHHHHPHHHPHPPXHHHPHPPXHPPXPXX`N11MethodTable16MethodDataObjectE", static c_NonVirtualSlotsOffsets = <same as static member of an already seen type>,
  static c_ModuleOverrideOffsets = <same as static member of an already seen type>,
  static c_OptionalMembersStartOffsets = <same as static member of an already seen type>, m_dwFlags = 2148139012, m_BaseSize = 24, m_wFlags2 = 11, m_wToken = 0,
  m_wNumVirtuals = 24, m_wNumInterfaces = 6, debug_m_szClassName = 0xff7d549018 "Int32Wrapper[]", m_pParentMethodTable = {static isRelative = false,
    m_addr = 1097312007520}, m_pLoaderModule = {static isRelative = false, m_ptr = 1097316892360}, m_pWriteableData = {static isRelative = false, m_ptr = 1097319354336},
  {m_pEEClass = {static isRelative = false, m_ptr = 1097319353744}, m_pCanonMT = {static isRelative = false, m_ptr = 1097319353744}}, {m_pPerInstInfo = {
      static isRelative = false, m_ptr = 1097319345880}, m_ElementTypeHnd = 1097319345880, m_pMultipurposeSlot1 = 1097319345880}, {m_pInterfaceMap = {
      static isRelative = false, m_ptr = 1097319354032}, m_pMultipurposeSlot2 = 1097319354032}}
(gdb) p ((MethodTable*)0xff7d548e40)->IsArray()
$7 = 1                        
(gdb) p pMethodTable2->GetNumInstanceFieldBytes()
$6 = 24                       
(gdb)
```

The above test case is `src/tests/Interop/StructMarshalling/PInvoke/MarshalStructAsLayoutSeq.cs` and the struct is https://github.com/dotnet/runtime/pull/62885#discussion_r819261350

Reply:
`GetNumInstanceFieldBytes() == 24` is the size of object header + methodtable + array size field. It is not related in any way to how many elements are in the array. If this condition makes `MarshalStructAsLayoutSeq` test pass, it is just a happy accident. It is going to fail if the actual size of the array is different.

This condition is not right.

Reply:
> `GetNumInstanceFieldBytes() == 24` is the size of object header + methodtable + array size field. It is not related in any way to how many elements are in the array. If this condition makes `MarshalStructAsLayoutSeq` test pass, it is just a happy accident. It is going to fail if the actual size of the array is different.
> 
> This condition is not right.

Thanks very much.
I will modify it.

Reply:
@jkotas 
Thanks very much for your review and guidance.